toolchain/parse/handle_decl_scope_loop.cpp

```diff
@@ -2,12 +2,54 @@
 // Exceptions. See /LICENSE for license information.
 // SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
+#include "toolchain/lex/token_kind.h"
 #include "toolchain/parse/context.h"
+#include "toolchain/parse/node_kind.h"
 
 namespace Carbon::Parse {
 
-static auto OutputInvalidParseSubtree(Context& context, int32_t subtree_start)
-    -> void {
+// Handles positions which are end of scope and packaging declarations. Returns
+// true when either applies. When the position is neither, returns false, and
+// may still update packaging state.
+static auto TryHandleEndOrPackagingDecl(Context& context) -> bool {
+  switch (context.PositionKind()) {
+    case Lex::TokenKind::CloseCurlyBrace:
+    case Lex::TokenKind::FileEnd: {
+      // This is the end of the scope, so the loop state ends.
+      context.PopAndDiscardState();
+      return true;
+    }
+    // `import`, `library`, and `package` manage their packaging state.
+    case Lex::TokenKind::Import: {
+      context.PushState(State::Import);
+      return true;
+    }
+    case Lex::TokenKind::Library: {
+      context.PushState(State::Library);
+      return true;
+    }
+    case Lex::TokenKind::Package: {
+      context.PushState(State::Package);
+      return true;
+    }
+    default: {
```

Comment:
Optional: to reduce indentation, consider `break;`ing here and putting the `default` case after the loop.